Inspection tools

JPXKQX opened the pull request:
This PR includes new features for inspecting the graphs:

- Plots: Interactive plots of the graphs. Distribution plots of edge & node attributes & isolated nodes.
- Graph description in the console
- CLI entry point: anemoi-graphs inspect graph.pt ...

Fixes:
- Argument `path` in GraphCreator is no longer mandatory.

This work extends the plotting functions created initially by @mishooax, @ssmmnn11 and the rest of the AIFS team.

FussyDuck commented:
[![CLA assistant check](https://bol-claassistant.ecmwf.int:443/pull/badge/not_signed)](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=22) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=22) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: JPXKQX<br/>:x: theissenhelen<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-graphs?pullRequest=22) it.</sub>

anaprietonem commented on src/anemoi/graphs/inspector.py:
```diff
@@ -0,0 +1,200 @@
+        assert self.output_path.is_dir(), f"Path {self.output_path} is not a directory."
+        assert os.access(self.output_path, os.W_OK), f"Path {self.output_path} is not writable."
+
+    def run_all(self):
```
I would propose to call this method 
`def inspect(self)`
to be more consistent with the description tool

anaprietonem commented on src/anemoi/graphs/inspector.py:
```diff
@@ -0,0 +1,200 @@
+        print()
+
+
+class GraphInspectorTool:
```
I would try to name `GraphInspectorTool` and `GraphDescription` following the same way so either:
- GraphInspectorTool and GraphDescriptorTool
- GraphInspection and GraphDescription

(as possible suggestions)

theissenhelen replied:
How about GraphInspector and GraphDescriptor?

anaprietonem replied:
I like it!

anaprietonem reviewed (commented):
Great work Mario! I have left a few early comments. Will keep reviewing the plotting functions tomorrow and would also like to test the code running the example you shared!

anaprietonem commented:
Few comments from testing this branch (recipe file used to create and inspect the graph attached (yaml can't be uploaded to git)
[recipe.txt](https://github.com/user-attachments/files/16597701/recipe.txt)

`1. anemoi-create recipe.yaml graph.pt`    
I found the output tables nice and informative, but I will try to add a table tittle or some additional prints to inform first table is the node summary and second one is about edge summary. And (this might be a naive question) but for the edges summary, the attributes dimension is reported as 3 but then the list of attributes just show two values?

![Screenshot 2024-08-13 at 12 07 50](https://github.com/user-attachments/assets/a1bbada9-fc31-4e30-8918-741b546289af)

I also test the option to deactivate description/overwriting. How do we plan to document this? maybe an option could be to use the README.md. I am asking this since, I found it not really intuitive that when passing the --description flag then the description is deactivated 😅
![Screenshot 2024-08-13 at 12 10 39](https://github.com/user-attachments/assets/29b112c8-d757-4b0f-927b-67e22eaf6a17)

 `2. anemoi-graphs inspect graph.pt`  
 
 
- Very nice command, plots get generated really fast 🚀
- The command allows to activate/deactivate --show_attribute_distributions, is there a similar option to be used with --show-nodes? (otherwise by default is always False). If I change to True (`show_nodes: Optional[bool] = True`,) then I hit the following error:
![Screenshot 2024-08-13 at 12 19 16](https://github.com/user-attachments/assets/35c73538-f56e-482f-b31a-a9bf84ec4aea)



------------------
To be able to run those commands I noticed that:
-   function number doesn't not exist in anemoi-utils [hummanize](https://github.com/ecmwf/anemoi-utils/blob/968267f40e023908733d0633a24d48a9e1dbaf59/src/anemoi/utils/humanize.py) script
-  Plotly is missing as dependency (required for plot generation)
- Need to rebase/bring changes from https://github.com/ecmwf/anemoi-graphs/pull/28/files

anaprietonem commented on src/anemoi/graphs/plotting/interactive_html.py:
```diff
@@ -0,0 +1,234 @@
+    isolated_nodes = compute_isolated_nodes(graph)
+
+    if len(isolated_nodes) == 0:
+        LOGGER.warning("No orphan nodes found.")
```
To be consistent, I would replace this with 'isolated nodes' rather than orphans

anaprietonem commented on src/anemoi/graphs/plotting/interactive_html.py:
```diff
@@ -0,0 +1,234 @@
+    node_traces = {}
+    for node_attr in node_attrs:
+        node_attr_values = graph[nodes_name][node_attr].float().numpy()
+        if node_attr_values.ndim > 1 and node_attr_values.shape[1] > 1:
```
what is this check for?

JPXKQX replied:
Comment added. Only 1D attributes are represented in the interactive plots.

anaprietonem commented on src/anemoi/graphs/plotting/interactive_html.py:
```diff
@@ -0,0 +1,234 @@
+    node_latitudes = np.rad2deg(graph[nodes_name].x[:, 0].numpy())
+    node_longitudes = np.rad2deg(graph[nodes_name].x[:, 1].numpy())
+    node_attrs = graph[nodes_name].node_attrs()
+    node_attrs.remove("x")
```
maybe a line to explain why are we dropping x?

anaprietonem commented on src/anemoi/graphs/plotting/displots.py:
```diff
@@ -0,0 +1,84 @@
+LOGGER = logging.getLogger(__name__)
+
+
+def plot_dist_node_attributes(graph: HeteroData, out_file: Optional[Union[str, Path]] = None) -> None:
```
Those two functions are quite similar and hence there is a bit of repeated code across them. I think it could be easier to define base function to plot_distributions that could be reused to plot both nodes and edges distributions.

anaprietonem commented on src/anemoi/graphs/plotting/interactive_html.py:
```diff
@@ -0,0 +1,234 @@
+    out_file : str, optional
+        Name of the file to save the plot. Default is None.
+    """
+    node_latitudes = np.rad2deg(graph[nodes_name].x[:, 0].numpy())
```
In the other functions we make use of node_list function to extract the lat/lons, why we don't do it in this case?

JPXKQX replied:
Good point! Updated to use node_list() function.

anaprietonem commented on src/anemoi/graphs/plotting/prepare.py:
```diff
@@ -0,0 +1,179 @@
+    return isolated_nodes
+
+
+def _get_node_attribute_dims(graph: HeteroData) -> dict[str, int]:
```
Similarly between _get_node_attribute_dims and _get_edge_attribute_dims is there a better way to abstract this functionality to reduce the amount of repeated code?

theissenhelen replied:
I also believe these shouldn't be private functions

anaprietonem commented on src/anemoi/graphs/plotting/interactive_html.py:
```diff
@@ -0,0 +1,234 @@
+            "line_width": 2,
+        },
+    )
+    layout = go.Layout(
```
Happy to discuss this. But I think in the future it could be worth refactoring this script in a similar fashion as we have in aifs-mono the callbacks.py/plots.py, with functions like _init_plot_settings or base_plot class where we can abstract the common functionality between all the plotting functions and then simply have each function to define the specifics of that method

anaprietonem commented:
I have left some comments in the PR, @JPXKQX pls feel free to reach out if something is not clear! Great work and really nice feature! I had a few general questions/future ideas:
- Do we have in mind adding inspection functionality for the edge attributes? In the tables I can see we have area_weights for the nodes, and edge length/directions so was wondering if we could also inform about the shape or visually inspect them.
- When generating the plots with 'plot_interactive_subgraph', the slider shows the level of connectivity. I was wondering if we could log some information about the values recorded so we know the distribution of the node connections for hidden->hidden, data->hidden and hidden-> data. It could be either some logs prints or a plot of the adjancecy matrix (or that's too big to plot 😅)?

JPXKQX commented:
> I have left some comments in the PR, @JPXKQX pls feel free to reach out if something is not clear! Great work and really nice feature! I had a few general questions/future ideas:
> 
> * Do we have in mind adding inspection functionality for the edge attributes? In the tables I can see we have area_weights for the nodes, and edge length/directions so was wondering if we could also inform about the shape or visually inspect them.
> * When generating the plots with 'plot_interactive_subgraph', the slider shows the level of connectivity. I was wondering if we could log some information about the values recorded so we know the distribution of the node connections for hidden->hidden, data->hidden and hidden-> data. It could be either some logs prints or a plot of the adjancecy matrix (or that's too big to plot 😅)?

* Currently, the user has the option of plotting the distribution of these edge attributes. There are 2 main problems with including them in the interactive plots: it scales worse (as there are more number of edges than nodes) and the resulting plot may not be as visually pleasing as a lot of edges may cross.
* That is a great idea! We can include both print statements in GraphDescriptor and distribution plots in GraphInspector.

JPXKQX commented:
Thanks for all the feedback @anaprietonem ! We have tried to address all the feedback below: 

>I found the output tables nice and informative, but I will try to add a table tittle or some additional prints to inform first table is the node summary and second one is about edge summary. And (this might be a naive question) but for the edges summary, the attributes dimension is reported as 3 but then the list of attributes just show two values?

The table titles have been added to the logged report. The attribute dimension is 3 because the edge_dirs is 2D. We could split the attribute dimension by attribute, but that might add too much information to the report, what do you think? 

>I also test the option to deactivate description/overwriting. How do we plan to document this? maybe an option could be to use the README.md. I am asking this since, I found it not really intuitive that when passing the --description flag then the description is deactivated 😅

Good point! It is now fixed. The user can pass "--description" (default) to leave the option activated or "--no-description" to deactivate it. This is documented in 'anemoi-graphs inspect --help' and will also be included in the official documentation when PR #19 is merged.

>The command allows to activate/deactivate --show_attribute_distributions, is there a similar option to be used with --show-nodes? (otherwise by default is always False)

Added new options "--show_nodes" and "--no-show_nodes" (default) to control the interactive node plots.

>If I change to True (show_nodes: Optional[bool] = True,) then I hit the following error:

I cannot reproduce this error now.

PD. An additional table has been added for printing the attribute statistics

anaprietonem commented:
Great thanks for the effort @theissenhelen and @JPXKQX! tested the updated code and:
- New cli arguments look good
- I like the new plots to inform about connectivity and table for statistics summary for attribute distributions
- About the dimensionality and plotting of the edge directions, I would maybe try then to indicate in the edges summary table under 'Attributes' columns something like: edge_length(1), edge_dirs(2) or either:  edge_dirs_0,edge_dirs_1, edge_length (same names as the one showed in the 'edge attributes distribution figures'). And thanks for clarifying the point about visually plotting that information, I now understand and agree with that. 
- Regarding the error I still see it, so let me see if I can try to explain how to reproduce it, the exact command is:
`anemoi-graphs inspect --show_nodes graph.pt output_plots`
Since by default we have --no_show_nodes I wanted to test setting this to True.

![Screenshot 2024-08-21 at 10 30 20](https://github.com/user-attachments/assets/90600e7c-34b7-4d38-b92e-cefe6c95cdb6)


I can share via Teams the exact recime.yaml and graph.pt to better reproduce this!

anaprietonem commented:
Discussed this on Teams, and the issues I reported is now fixed. Great work both of you, my suggestion and comments are now solved, so happy to approve this PR!
